### DISCUSSION TITLE
Should order of Groups matter? Should Groups depend on Files?

### DISCUSSION THREAD

**Discussion by shwarcu** [LironEr/bundlemon#96] [Q&A] [answered]
Hi, I just noticed that order of defining groups changes result of bundlemon

Given such config 

```json
{
  "baseDir": "./dist",
  "reportOutput": ["json"],
  "defaultCompression": "none",
  "groups": [
    {
      "path": "vendors~*.js"
    },
    { "path": "*.js" }
  ],
  "files": [...]
}

```

Result is following
```
Groups:
  [PASS] *.js: 5MB
  [PASS] vendors~*.js: 1MB
```

But if 'vendors' group is defined after 'js' group, then it's ignored:
```json
{
  "baseDir": "./dist",
  "reportOutput": ["json"],
  "defaultCompression": "none",
  "groups": [
    { "path": "*.js" },
    {
      "path": "vendors~*.js"
    }
  ],
  "files": [...]
}

```
gives

```
Groups:
  [PASS] *.js: 6MB (vendors group is ignored)
```

Therefore I have 2 questions
1. Should order of defining groups matter? If yes, maybe documentation should mention that and be a bit clearer how grouping works
2. Could Grouping work independently from `Files`? From what I understand at the moment Groups only include files that are defined by `Files` array.  If they work independently it would be a good change for a big project of a case where I would like to track some `Files` that are critical for the app, but use `Groups` in order to track everything that is produced by bundler.

I have a case, where dist folder contains about 100 small files + some files that I want to track.
I would like to track most important files with `Files` and then track everything with `Groups`.
But If I add everything (100+) files to `Files` array, it will generate huge report and I would prefer not going this way :(

**Comment by LironEr** [comment, marked as answer]
Hey,

1. Groups order shouldn't matter, this is a bug. I will release a fix in the next few minutes.
2. Groups work independently from files, you can have different patterns, for example:
```json
{
  "files": [
    {
      "path": "index.html"
    },
    {
      "path": "static/js/*.<hash>.chunk.js"
    }
  ],
  "groups": [
    {
      "path": "**/*.js"
    },
    {
      "path": "**/*.css"
    }
  ]
}
```

**Reply by shwarcu** [reply]
Thanks a lot for such a fast and clear reply with action @LironEr 😃 
In case if I don't want to have any files defined, should it be allowed to have empty array like this?: 

```json
{
  "files": [],
  "groups": [
    {
      "path": "**/*.js"
    },
    {
      "path": "**/*.css"
    }
  ]
}
```

When I tried such config I got an error. I can reproduce it later if needed.
But I didn't try to not-include `files` property at all yet

**Reply by LironEr** [reply]
Currently, it's not allowed, but I don't think of any reason why it should be like that.
I will change it that both `files` and `groups` are optional.

a workaround for now: 
```json
{
  "files": [
    {
      "path": "dont-find-me"
    }
  ],
  "groups": [
    {
      "path": "**/*.js"
    },
    {
      "path": "**/*.css"
    }
  ]
}
```

**Reply by shwarcu** [reply]
Thanks! :)
